### PR TITLE
[Merged by Bors] - feat: change `Subgroup` and `Submonoid` induction principles to work with `induction`

### PR DESCRIPTION
**Pull request by eric-wieser**
Induction principles have to be fully dependent in order to work with the `induction` tactic. This is usually a good thing anyway, since the dependent version is often more convenient to use, and avoids the caller having to fumble with generalizing over an existential.

This changes the following induction principles (and their additive versions):
* `Submonoid.closure_induction_{left,right}`
* `Subgroup.closure_induction_{left,right}`
* `Subgroup.closure_induction''` (no submonoid version exists)

Arguments to these lemmas have also been renamed to drop the `H`, as seems to be preferred for `induction` lemmas.

---

- [x] depends on: #9860

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/GroupTheory/Subgroup/Pointwise.lean [hunk 1]
```diff
@@ -74,24 +74,44 @@ theorem closure_toSubmonoid (S : Set G) :
...
-  Submonoid.closure_induction_right (key h) H1 fun x y hy => hy.elim (Hmul x y) fun hy hx =>
-    inv_inv y ▸ Hinv x y⁻¹ hy hx
+@[to_additive (attr := elab_as_elim)
+  "For additive subgroups generated by a single element, see the simpler
+  `zsmul_induction_right`."]
+theorem closure_induction_right {p : (x : G) → x ∈ closure s → Prop} (H1 : p 1 (one_mem _))
+    (mul_right : ∀ (x) hx, ∀ y (hy : y ∈ s), p x hx → p (x * y) (mul_mem hx (subset_closure hy)))
+    (mul_right_inv : ∀ (x) hx, ∀ y (hy : y ∈ s), p x hx →
+      p (x * y⁻¹) (mul_mem hx (inv_mem (subset_closure hy))))
+    {x : G} (h : x ∈ closure s) : p x h := by
```
Comment (adomani) [inline Mathlib/GroupTheory/Subgroup/Pointwise.lean:105]: Can't you also prove this one using the opposite product and recycle the `closure_induction_left` proof?

**Review by adomani — APPROVED** [review]
This is a great improvement, thanks!

**Comment by jcommelin** [comment]
LGTM, module @_adomani's comment.

bors d+

File: Mathlib/GroupTheory/Subgroup/Pointwise.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/GroupTheory/Subgroup/Pointwise.lean]: Sure, done

Comment (adomani) [reply Mathlib/GroupTheory/Subgroup/Pointwise.lean]: Very cool, thanks!

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**